### PR TITLE
LPD-5780 - functional test for COMMERCE-13162

### PR DESCRIPTION
**Pull request by wanderlast**
https://liferay.atlassian.net/browse/LPD-5780
https://liferay.atlassian.net/browse/COMMERCE-13162

Functional test for COMMERCE-13162. This is meant to check the following for a product ("Product A") with a potentially escapable character in it:
- The modal title when attempting to add a product relation to Product A.
- The name displayed for product A after adding it as a product relation to another product.

### CODE REVIEW

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/ActionsDropdown.tsx [hunk 1]
```diff
@@ -227,6 +227,7 @@ function ActionsDropdown({
 								'ml-1': !Liferay.FeatureFlags['LPS-193005'],
 							}
 						)}
+						data-qa-id="itemActionDropdown"
```
Comment (brianikim) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/ActionsDropdown.tsx:230]: Could you make sure if this is a duplicate of another data-qa-id? If it is, please double check whether there is possibility of both uses of data-qa-id being on the same page.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CreationMenu.tsx [hunk 2]
```diff
@@ -60,6 +60,7 @@ const DropDown = ({
 				<ClayButton
 					aria-label={Liferay.Language.get('new')}
 					className="nav-btn nav-btn-monospaced"
+					data-qa-id="creationMenuNewButton"
```
Comment (brianikim) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CreationMenu.tsx:63]: Could you make sure if this is a duplicate of another data-qa-id? If it is, please double check whether there is possibility of both uses of data-qa-id being on the same page.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CreationMenu.tsx [hunk 3]
```diff
@@ -131,6 +132,7 @@ function CreationMenu({
 						<ClayButton
 							aria-label={primaryItems[0].label}
 							className="nav-btn nav-btn-monospaced"
+							data-qa-id="creationMenuNewButton"
```
Comment (brianikim) [inline modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CreationMenu.tsx:135]: Could you make sure if this is a duplicate of another data-qa-id? If it is, please double check whether there is possibility of both uses of data-qa-id being on the same page.

File: modules/test/playwright/pages/commerce/commerceProductsPage.ts [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+    readonly newMenuButton: Locator;
+    readonly spareProductMenuButton: Locator;
+    readonly page: Page;
+
+    constructor(page: Page) {
+        
+        this.applicationsMenuPage = new ApplicationsMenuPage(page);
+        this.deleteMenuItem = page.getByRole('menuitem', {name: 'Delete'});
+        this.itemOptionMenu = page.getByTestId('itemActionDropdown');
+        this.modalAddButton = page.locator('modal-item-last').getByRole('button', {name: 'Add'});
```
Comment (brianikim) [inline modules/test/playwright/pages/commerce/commerceProductsPage.ts]: We need to use a more specific locator. testId if possible.

File: modules/test/playwright/pages/commerce/commerceProductsPage.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+        this.modalCloseButton = page.locator('modal-header').getByRole('button', {name: 'close'});
```
Comment (brianikim) [inline modules/test/playwright/pages/commerce/commerceProductsPage.ts]: We need to use a more specific locator. testId if possible.

File: modules/test/playwright/pages/commerce/commerceProductsPage.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+			await this.spareProductMenuButton.click();
+		}
+    }
+
+    async goto() {
+        await this.applicationsMenuPage.goToProducts();
+    }
+
+    async goToProductRelations() {
+        await this.page.getByRole('link', { name: 'Product Relations' }).click();
```
Comment (brianikim) [inline modules/test/playwright/pages/commerce/commerceProductsPage.ts]: Do not construct new locators in the method. These need to be declared in the constructor.

File: modules/test/playwright/pages/commerce/commerceProductsPage.ts [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    async goToSpecificProductMenu(productName : String) {
+        await this.page.getByRole('link', {name: productName}).click();
```
Comment (brianikim) [inline modules/test/playwright/pages/commerce/commerceProductsPage.ts]: Do not construct new locators in the method. These need to be declared in the constructor.

Comment (brianikim) [reply modules/test/playwright/pages/commerce/commerceProductsPage.ts]: We have existing examples of passing in parameters in the constructor implementations.

File: modules/test/playwright/tests/commerce/commerceProducts.spec.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    commerceProductsPage
+}) => {
+
+    const catalog = await apiHelpers.headlessCommerceAdminCatalog.postCatalog(
+		'Product Catalog'
+	);
+
+    const product1 = await apiHelpers.headlessCommerceAdminCatalog.postProduct(
+		catalog.id,
+		'\"Product1\"'
```
Comment (brianikim) [inline modules/test/playwright/tests/commerce/commerceProducts.spec.ts]: I think all of these post calls randomize the names by default. Do not provide hard-coded names if not required.

File: modules/test/playwright/tests/commerce/commerceProducts.spec.ts [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+    const product2 = await apiHelpers.headlessCommerceAdminCatalog.postProduct(
+        catalog.id,
+        'Product2'
+    );
+
+	//double-check data set
+	const returnedProduct = await apiHelpers.headlessCommerceAdminCatalog.getProduct(
+		product1.productId);
+
+	expect(returnedProduct.name).toBe('\"Product1\"');
```
Comment (brianikim) [inline modules/test/playwright/tests/commerce/commerceProducts.spec.ts]: Fix this accordingly as well.

File: modules/test/playwright/tests/commerce/commerceProducts.spec.ts [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+
+	//now check in the actual GUI
+    await commerceProductsPage.goto();
+
+	await commerceProductsPage.goToSpecificProductMenu(product1.name);
+
+	await commerceProductsPage.addSpareProductRelation();
+
+	await expect(
+		commerceProductsPage.getByRole(
```
Comment (brianikim) [inline modules/test/playwright/tests/commerce/commerceProducts.spec.ts]: No locators outside of page.ts

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/ActionsDropdown.tsx (hunk 1 above)
Comment (wanderlast) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/ActionsDropdown.tsx:230]: This is not a duplicate

File: modules/test/playwright/tests/commerce/commerceProducts.spec.ts [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+	await commerceProductsPage
+		.frameLocator('#modalIframe')
+		.getByRole('checkbox', {value: product1.productId})
+	.check();
+
+	await commerceProductsPage.modalAddButton.click();
+
+	await expect(
+		commerceProductsPage.getByRole('link', {name: product1.name}))
+	.toBeVisible();
```
Comment (brianikim) [inline modules/test/playwright/tests/commerce/commerceProducts.spec.ts]: Same here. Declare the locators all in page.ts

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CreationMenu.tsx (hunk 2 above)
Comment (wanderlast) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CreationMenu.tsx:63]: This is a duplicate of what you named it in CreationMenu.js. AFAIK you cannot have both implementations of CreationMenu on the same page.

File: modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CreationMenu.tsx (hunk 3 above)
Comment (wanderlast) [reply modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CreationMenu.tsx:135]: See above

File: modules/test/playwright/tests/commerce/commerceProducts.spec.ts (hunk 8 above)
Comment (wanderlast) [reply modules/test/playwright/tests/commerce/commerceProducts.spec.ts]: So the problem with purely randomizing the name is that we need to have a character of some kind in the name that would normally be escaped and would therefore appear incorrectly in the case of UI for this specific test case.

Comment (wanderlast) [reply modules/test/playwright/tests/commerce/commerceProducts.spec.ts]: Would it be preferred if I had it generate a randomized name and then add the character to be escaped? I just figured this was cleaner to hard-code so it's a bit more readable.

**Comment by wanderlast** [comment]
ci:test:sf

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by wanderlast** [comment]
ci:test:playwright

**Comment by brianikim** [comment]
Hello @wanderlast , please provide ticket information on the PR + link PR to the ticket.

**Comment by brianikim** [comment]
@wanderlast , you can forward to Stefano once you finished all the requested changes.

**Pull request closed without merging**